sanitizers/src/main/java/com/code_intelligence/jazzer/sanitizers/XpathInjection.kt

```diff
@@ -0,0 +1,95 @@
+// Copyright 2022 Code Intelligence GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.code_intelligence.jazzer.sanitizers
+
+import com.code_intelligence.jazzer.api.FuzzerSecurityIssueHigh
+import com.code_intelligence.jazzer.api.HookType
+import com.code_intelligence.jazzer.api.Jazzer
+import com.code_intelligence.jazzer.api.MethodHook
+import com.code_intelligence.jazzer.api.MethodHooks
+import java.lang.invoke.MethodHandle
+
+/**
+ * Detects Xpath injections.
+ *
+ * Untrusted input has to be escaped in such a way that queries remain valid otherwise an injection
+ * could be possible. This sanitizer guides the fuzzer to inject insecure characters. If an exception
+ * is raised during execution the fuzzer was able to inject an invalid pattern, otherwise all input
+ * was escaped correctly.
+
+ * False positives can be possible but should bugs in the code just not an injection. e.g.
+ * "If expression contains any XPathFunctions, they must be available via the XPathFunctionResolver. An XPathExpressionException will be thrown if the XPathFunction cannot be resovled with the XPathFunctionResolver."
+ * "The parameter item represents the context the XPath expression will be operated on. The type of the context is implementation-dependent. If the value is null, the operation must have no dependency on the context, otherwise an XPathExpressionException will be thrown."
+ * https://docs.oracle.com/en/java/javase/17/docs/api/java.xml/javax/xml/xpath/XPath.html
+ * 
+ * Checking the number of quotation marks to be uneven should indicate injection instead of a false positive 
```

Comment:
Could you explain what such a false positive would look like? Why are quotation marks relevant for this?

Reply:
The XPathExpressionException is kinda generall and can have many reasons. Like for example "javax.xml.xpath.XPathExpressionException: java.io.IOException: InputStreamReader is closed" Since this is hopefully expected and dealt with in the fuzzed app and also is no injection, that would be a false positive. To prevent them I thought of just checking the number of single and double quotes in the expression. From my understanding they should always be even, so if they are not it is an possible injection. The alternative is that the app adds an additional quote somewhere which throws the exception. While that would not be an injection it still is a bug and should only happen rarely.

I checked some small test examples and it works correctly for them. Naturally there can be cases that I have not considered. Since I started reading about Xpath just for this sanitizer so I would not call myself an expert ^^